CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

IntegralProgrammer opened the pull request:
This PR builds on top of the CARDS-1359 branch and provides the necessary UI and reverse proxy changes so that a user may login to CARDS either as a native Sling user or through a SAML Identity Provider (IdP). If a SAML IdP is configured, a two-step login process will be presented to the user where, the first step involves entering a _username_. If the _username_ is a native Sling user, the second step will be a password prompt for the given username. If the _username_ indicates a SAML login from a configured IdP (eg. `myuser@keycloakdemo`) the user will be redirected to login at the IdP login screen.

To test:

- Build this (`CARDS-1374`) branch.
- Follow the instructions in `Utilities/Administration/SAML/SAML_SETUP.md` to setup a local Keycloak Docker container as an IdP.
- Ensure that, when visiting `http://localhost:9090/`, you are able to login both as a native Apache Sling user (`admin`:`admin`) and as the `myuser@keycloakdemo` SAML user.
- After logging out as the `admin` user, ensure that no subsequent requests to `http://localhost:9090` contain the `sling.formauth` cookie.
- After logging out as the `myuser@keycloakdemo` SAML user, log back in as `myuser@keycloakdemo`. If the Single Sign On (SSO) session has not expired, you will not be prompted by Keycloak for the `myuser@keycloakdemo` _username_ and _password_ but you should still briefly see the Keycloak authentication redirection screen.
- Ensure that `fetchWithReLogin` (popup login dialog) functionality works properly for both a native Apache Sling user as well as a SAML authenticated user.
- Ensure that the _Security Tests_ in `Utilities/Administration/SAML/SAML_SETUP.md` also pass.
- Ensure that if `start_cards.sh` is invoked without the `--keycloak_demo` flag, that the regular one-step login form is presented to the user.

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Basic two-step login process setup

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

If the user has specified a remote domain (with the @ symbol), then check if the remote domain is listed under /apps/cards/SAMLDomains in the JCR repository. If it is not, display a failed login error

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

If the user has specified a valid remote domain, open a browser-native popup window to that domain's login page

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Auto-focus the Password input when it is displayed and display the selected username in the label

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Fix indentation

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add a Back button on the password entry phase so that the user can go backwards if they mistyped their username

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Cause the popup window for the identity provider login to open in the center of the screen

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Display more meaningful error messages

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Only operate in two-step mode if the /apps/cards/SAMLDomains JCR directory exists

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add a --keycloak_demo flag for start_cards.sh which, when present, allows users to login with an @keycloakdemo email address which will authenticate with a Keycloak server on localhost:8484 under the realm "myrealm"

This is to be used primarily for development and testing purposes

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add the /fetch_requires_saml_login.html endpoint so that SAML logins can be used when fetchWithReLogin() causes a login dialog to appear

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

If the login screen is shown as the main window and a user provides a federated login ID (eg. an @uhn.ca email address), redirect the browser window to the SAML login screen

If the login screen is shown as a dialog window and a user provides a federated login ID, open the SAML login screen in a pop-up browser window

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add the keycloak_headermod_http_proxy.js reverse proxy for testing keycloak_demo without requiring a complete Docker Compose setup involving an Apache HTTP reverse proxy container

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

keycloak_headermod_http_proxy.js reverse proxy now instructs the client to delete the sling.formauth cookie upon a HTTP 302 redirect issued by /system/sling/logout or /system/console/logout

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

add_saml_sp_config.py utility program now uses http://localhost:9090/ as the default value for --saml2LogoutURL instead of http://localhost:8484/

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Update README.md with instructions involving the @keycloakdemo testing IdP and the keycloak_headermod_http_proxy testing reverse proxy

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Replace hard-coded 0.9-SNAPSHOT version identifier in start_cards.sh with CARDS_VERSION constant

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add the reverse proxy rules for SAML to compose-cluster/proxy/http_000-default.conf

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Move the reverse proxy rules for SAML from compose-cluster/proxy/http_000-default.conf to compose-cluster/proxy/http_saml_000-default.conf

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Create a SSL protected version of http_saml_000-default.conf - https_saml_000-default.conf

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

When a CARDS Docker container created with SAML_AUTH_ENABLED==true, make the cards-fetch-requires-saml-login feature present and activated

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Enable generate_compose_yaml.py to configure the reverse proxy container for SAML integration

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add a setup_saml_docker.sh script for testing SAML authentication with a local Keycloak instance when CARDS is deployed with Docker Compose

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Fix https_saml_000-default.conf - redirect to CARDS login page accessed via HTTPS not HTTP

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add a Keycloak realm JSON file for a CARDS instance available at https://localhost/

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add the setup_saml_docker_ssl.sh script for configuring the local CARDS instance to authenticate users via SAML against the local Keycloak instance when the CARDS instance is made available on https://localhost/

IntegralProgrammer commented:
## Testing Instructions for Docker parts for CARDS-1374

- After performing the tests described at the beginning of this PR based on the instructions in `Utilities/Administration/SAML/SAML_SETUP.md`, repeat the same tests except this time, use Docker Compose to provision the CARDS deployment:
  - `cd compose-cluster`
  - `python3 generate_compose_yaml.py --cards_project cards4proms --oak_filesystem --saml --saml_idp_destination http://localhost:8484/auth/realms/myrealm/protocol/saml --server_address localhost:8080 --composum`
  - Edit the file `docker-compose.yml` and under `services`/`cardsinitial`/`environment`, add the entry `- ADDITIONAL_SLING_FEATURES=mvn:io.uhndata.cards/cards-keycloakdemo-saml-support/0.9-SNAPSHOT/slingosgifeature`
  - Copy the file `samlKeystore.p12` from the `Utilities/Administration/SAML` directory into the `compose-cluster` directory.
  - Ensure that you have a `cards/cards` Docker image built from this `CARDS-1374` branch. If your image is not self-contained (eg. not built with `docker.buildArg.build_jars=true`), edit the `docker-compose.yml` file so that your `~/.m2` directory is mounted (read-only) to the `cardsinitial` container's `/root/.m2` directory.
  - `docker-compose build`
  - `docker-compose up -d`
  - After CARDS has started, in the `Utilities/Administration/SAML` directory, run the command `./setup_saml_docker.sh`
  - Complete all the tests described in `Utilities/Administration/SAML/SAML_SETUP.md`
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - `./cleanup.sh`

- Verify that a normal (ie. without SAML) HTTP deployment of CARDS still works properly:
  - `python3 generate_compose_yaml.py --cards_project cards4proms --oak_filesystem`
  - `docker-compose build`
  - `docker-compose up -d`
  - Visit `http://localhost:8080` and ensure that everything works as expected
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - `./cleanup.sh`

- Create a self-signed SSL configuration
  - Within the `compose-cluster` directory, `mkdir SSL_CONFIG`
  - `openssl req -newkey rsa:4096 -nodes -keyout SSL_CONFIG/certificatekey.key -x509 -days 365 -out SSL_CONFIG/certificate.crt`
  - `cp SSL_CONFIG/certificate.crt SSL_CONFIG/certificatechain.crt`

- Verify that a normal (ie. without SAML) HTTPS deployment of CARDS still works properly:
  - `python3 generate_compose_yaml.py --cards_project cards4proms --oak_filesystem --ssl_proxy`
  - `docker-compose build`
  - `docker-compose up -d`
  - Visit `https://localhost`, accept the self-signed certificate, and ensure that everything works as expected
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - `./cleanup.sh`

- Verify that a SAML-enabled HTTPS deployment of CARDS works properly:
  - Follow the steps in`Utilities/Administration/SAML/SAML_SETUP.md` to setup the Keycloak server, except this time, use the file `Utilities/Adminstration/SAML/test_realm_ssl.json` instead of `Utilities/Administration/SAML/test_realm.json`.
  - `cd compose-cluster`
  - `python3 generate_compose_yaml.py --cards_project cards4proms --oak_filesystem --saml --saml_idp_destination http://localhost:8484/auth/realms/myrealm/protocol/saml --server_address localhost --composum --ssl_proxy`
  - Edit the file `docker-compose.yml` and under `services`/`cardsinitial`/`environment`, add the entry `- ADDITIONAL_SLING_FEATURES=mvn:io.uhndata.cards/cards-keycloakdemo-saml-support/0.9-SNAPSHOT/slingosgifeature`
  - Copy the file `samlKeystore.p12` from the `Utilities/Administration/SAML` directory into the `compose-cluster` directory.
  - Ensure that you have a `cards/cards` Docker image built from this `CARDS-1374` branch. If your image is not self-contained (eg. not built with `docker.buildArg.build_jars=true`), edit the `docker-compose.yml` file so that your `~/.m2` directory is mounted (read-only) to the `cardsinitial` container's `/root/.m2` directory.
  - `docker-compose build`
  - `docker-compose up -d`
  - After CARDS has started, in the `Utilities/Administration/SAML` directory, run the command `./setup_saml_docker_ssl.sh`
  - Complete all the tests described in `Utilities/Administration/SAML/SAML_SETUP.md`
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - `./cleanup.sh`

acrowthe commented:
Testing without Docker: Working on docker testing to reproduce concerns

Usability concern: When the two step login process is enabled (running with --keycloak_demo), pressing enter on the "Username or email address" field leads to an "Invalid username or password" error. I'd expect hitting enter to have the effect of hitting "next" and proceeding to either the cards password prompt or the keycloak login.
![image](https://user-images.githubusercontent.com/20056751/145843790-4e6c3bff-d706-47ff-a6c7-2c5065628724.png)

Usability concern: If the user enters their keycloak username in our username field and is then redirected to the SAML login page, I'd expect the username to be filled in with whatever they entered in the CARDS username field. Is this possible/easy to do, or is it going to be different for each SAML provider?
![image](https://user-images.githubusercontent.com/20056751/145844244-2a761c07-f2cf-4c06-9755-8107540f7f29.png)
![image](https://user-images.githubusercontent.com/20056751/145844304-a0a9543a-2a98-4a6c-84ab-2e0e32a0923a.png)

Testing the fetch_with_relogin:
 - Works with a local CARDS user (though it also has the "enter -> error" usability concern)
 - Does not work consistently with a SAML user when the same browser experiences fetch with relogin with both valid and invalid SSO sessions
   Reproduction case 1 (valid SSO -> invalid SSO):
   - Open a new incognito window to the keycloak admin page, sign in as admin and make sure myuser has no active sessions (Users -> Myuser -> Sessions -> Log out all sessions)
   - Open a new incognito window to localhost:9090 and sign in as myuser
   - Start creating a new subject (+ ->Patient or related entry -> next -> put in a name)
   - In a new tab in that browser window, navigate to localhost:9090 and sign out (M in top right -> sign out)
   - Return to the first tab and hit create, which should pull up the fetch with relogin window
   - Put in `myuser@keycloakdemo` and hit next. Popup should appear briefly, fetch with relogin should close and subject should be created
   - Start creating a new patient, pause at the same point
   - In the second tab, navigate back to localhost:9090 and sign out again
   - In the keycloak incognito window, use that admin account to sign out the myuser sessions again
   - In the cards window tab with the patient form ready to create, hit create to pull up fetch with relogin
   - Put in `myuser@keycloakdemo` and hit next. Popup appears briefly then closes with no user input, fetch with relogin stays open and patient does not create
 
   Reproduction case 2 (invalid SSO -> valid SSO):
   - Open a new incognito window to the keycloak admin page, sign in as admin and make sure myuser has no active sessions
   - Open a new incognito window to localhost:9090 and sign in as myuser
   - Start creating a new subject
   - In a new tab in that browser window, navigate to localhost:9090 and sign out
   - In the keycloak incognito window, use that admin account to sign out the myuser sessions
   - Return to the first cards window/tab and hit create, which should pull up the fetch with relogin window
   - Put in `myuser@keycloakdemo` and hit next. Popup should appear asking for username and password. Sign in as myuser, popup and fetch with relogin should close and patient should 
   - Start creating a new patient, pause at the same point
   - In the second tab, navigate back to localhost:9090 and sign out again
   - In the cards window tab with the patient form ready to create, hit create to pull up fetch with relogin
   - Put in `myuser@keycloakdemo` and hit next. Popup appears briefly then closes with no user input, fetch with relogin stays open and patient does not create

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

If we are in two-step login mode and are on the username entry phase, pressing the Enter key should have the same effect as clicking the Next button, instead of submitting the login form as was the previous behavior of this component

IntegralProgrammer commented:
> Usability concern: When the two step login process is enabled (running with --keycloak_demo), pressing enter on the "Username or email address" field leads to an "Invalid username or password" error. I'd expect hitting enter to have the effect of hitting "next" and proceeding to either the cards password prompt or the keycloak login.

Implemented this fix in `ade181f`. Now, if the user is on the username entry step of the two-step login, pressing Enter will cause the two-step login process to advance to the password entry step as would be expected by most users.

IntegralProgrammer commented:
> Usability concern: If the user enters their keycloak username in our username field and is then redirected to the SAML login page, I'd expect the username to be filled in with whatever they entered in the CARDS username field. Is this possible/easy to do, or is it going to be different for each SAML provider?

Unfortunately there is no one-size-fits-all solution to do this. I looked to see if this would be possible with Keycloak and it didn't look like there was any URL parameter that would do this. Thankfully, for the _UHN AD FS_ SAML IdP it does look possible through the use of the `username` URL parameter (eg. `https://fs.uhn.ca/adfs/ls/?wa=signin1.0&wtrealm=urn:federation:MicrosoftOnline&username=john.smith@uhn.ca`).

So, to get that to work in a production environment, we would need the `loginForm` component to pass the entered username through to the SAML IdP URL that it opens as a URL parameter. How that parameter would be passed would most likely be configured by the `/apps/cards/SAMLDomains/IdpDescriptor` node.

It's a good idea although I'm not sure if it's within the scope of CARDS-1374. @marta- ?

marta- commented:
> It's a good idea although I'm not sure if it's within the scope of CARDS-1374. @marta- ?

Agreed, we should do it, but it seems to be more related to CARDS-1360.

OrdiNeu commented:
Bug found:
1. Login as `myuser@keycloakdemo` at `localhost:9090/login`
2. Sign out, get redirected to `localhost:9090/login`
3. Sign in as literally anybody `@keycloakdemo`
4. You will automatically redirect through keycloak and become logged in as `myuser`

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Revise setup_saml_docker.sh and setup_saml_docker_ssl.sh and remove the calls to the (now removed) Python programs which perform the configurations now done by the base.json feature file

acrowthe reviewed (approved):
Tested, everything works. 

Some minor config changes needed to handle Docker + HTTPS + SAML given a locally hosted `http:` SAML instance:
- Needed to remove one of the two `Header edit Location`s from `https_saml_000-default.conf` (the root one, not the one in `<Location "/">`. This was required or keycloak would be accessed via `https:` and would fail
- Needed to edit a lot of the `http:` in `add_saml_sp_config.py` to `https:` (skipped the `saml2IDDPestination` arg and `CARDS_URL`). Required due to the proxy config change.
- Worked around an issue where keycloak was redirecting to `http:localhost` instead of `https`, again due to the proxy config change.

As production instances will have an `https:` SAML instance, these workarounds should not be required.
